projects/deref-patterns.md

```diff
@@ -0,0 +1,45 @@
+# Proposal
+
+Allow pattern matching through types that impl `Deref` or `DerefMut`. 
+
+## Summary and problem statement
+
+Currently in rust, matching is blocked by bounderies like smart pointers, containers, and some wrappers. 
+To solve this problem you would need to either use if let guards (unstable), or nested match/if-let. 
+The former is limited to one such level, and the latter can become excessive for deeply nested types. 
+To solve this, it is proposed propose that "deref patterns" be added, to allow for such matching to be performed.
+
+An exception to the above problem, is that `Box<T>` can be matched with `feature(box_patterns)`. 
+However, this is magic behaviour of box, and a goal of this project is to remove that magic. 
+
+The proposed solution has a number of unanswered questions, including the syntax for patterns, whether or not to limit to standard library types,
+ and how to allow exhaustive patterns soundly if not limited to standard library types. 
+
+
+## Motivation, use-cases, and solution sketches
+
+Recursive types necessarily include smart pointers, even when you could normally match through them.
+For example, in a work in progress proc-macro to support restricted variadic generics, the parser needed to match "fold expressions", which take the form `(<pattern> <op> ...)`. With deref patterns, this could be implemented using `Expr::Paren(ParenExpr{expr: Expr::Binary(ExprBinary{ left, op, right: Expr::Verbaitum(t), ..}), ..})`. However, this is currently not possible, and required nested matches.  
```

Comment:
```suggestion
For example, in a work in progress proc-macro to support restricted variadic generics, the parser needed to match "fold expressions", which take the form `(<pattern> <op> ...)`. With deref patterns, this could be implemented using `Expr::Paren(ParenExpr{expr: Expr::Binary(ExprBinary{ left, op, right: Expr::Verbatim(t), ..}), ..})`. However, this is currently not possible, and required nested matches.  
```